File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def oneNum(self, nums, target, first):
        left = first + 1
        right = len(nums) - 1
        closestSum = nums[first] + nums[left] + nums[right]

        while left < right:
            currentSum = nums[first] + nums[left] + nums[right]

            if currentSum == target:
                return target

            if abs(currentSum - target) < abs(closestSum - target):
                closestSum = currentSum

            if currentSum < target:
                left += 1
            else:
                right -= 1

        return closestSum

    def threeSumClosest(self, nums, target):
        n = len(nums)
        if n < 3:
            return 0
        nums.sort()
        m = nums[0] + nums[1] + nums[2]

        for i in range(n - 2):
            temp = self.oneNum(nums, target, i)
            if abs(temp - target) < abs(m - target):
                m = temp

        return m

File: 0016-3sum-closest/0016-3sum-closest.cpp
#include <vector>
#include <algorithm>
#include <cstdlib>
using namespace std;

class Solution {
public:
    int oneNum(vector<int>& nums, int target, int first) {
        int left = first + 1;
        int right = nums.size() - 1;
        int closestSum = nums[first] + nums[left] + nums[right];

        while (left < right) {
            int currentSum = nums[first] + nums[left] + nums[right];

            if (currentSum == target) return target;

            if (abs(currentSum - target) < abs(closestSum - target)) {
                closestSum = currentSum;
            }

            if (currentSum < target) {
                left++;
            } else {
                right--;
            }
        }

        return closestSum;
    }

    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        if (n < 3) return 0;
        sort(nums.begin(), nums.end());
        int m = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < n - 2; i++) {
            int temp = oneNum(nums, target, i);
            if (abs(temp - target) < abs(m - target)) {
                m = temp;
            }
        }

        return m;
    }
};
